common/configurable/src/main/java/io/helidon/common/configurable/ThreadPool.java

```diff
@@ -498,31 +505,91 @@ void setPool(ThreadPool pool) {
 
         @Override
         public boolean offer(Runnable task) {
-            return enqueue(task);
+            if (!semaphoreWrite.tryAcquire()) {
+               return false;
+            }
+            try {
+                enqueue(task);
+            } catch (Exception e) {
+                semaphoreWrite.release();
+                return false;
+            }
+            semaphoreRead.release();
+            return true;
+        }
+
+        @Override
+        public Runnable poll() {
+            if (!semaphoreRead.tryAcquire()) {
+               return null;
+            }
+            semaphoreWrite.release();
+            return super.poll();
+        }
+
+        @Override
+        public boolean offer(Runnable task, long timeout, TimeUnit tu) throws InterruptedException {
+           if (!semaphoreWrite.tryAcquire(timeout, tu)) {
+              return false;
+           }
+           try {
+               enqueue(task);
+           } catch (Exception e) {
+               semaphoreWrite.release();
+               return false;
+           }
+           semaphoreRead.release();
+           return true;
+        }
+
+        @Override
+        public Runnable poll(long timeout, TimeUnit tu) throws InterruptedException {
+           if (!semaphoreRead.tryAcquire(timeout, tu)) {
+              return null;
+           }
+           semaphoreWrite.release();
+           return super.poll();
+        }
+
+        @Override
+        public void put(Runnable task) throws InterruptedException {
+           semaphoreWrite.acquire();
+           try {
+               enqueue(task);
+           } catch (Exception e) {
+               semaphoreWrite.release();
+               throw e;
```

Comment:
This probably doesn't compile.

I would prefer a `try...finally` inside `enqueue`. The reason is that `enqueue` knows where the underlying queue has been touched successfully, and where it wasn't (anything happened after `offer` returned successfully). So if it was touched successfully, should `semaphoreRead.release`, otherwise should `semaphoreWrite.release`.

```
private boolean enqueue(...) {
   boolean offered = false;
   try {
      offered = super.offer(...);
      ...
   } finally {
      (offered? semaphoreRead: semaphoreWrite).release();
   }
}
```

Reply:
> This probably doesn't compile.

 This is fine. If it can be thrown by the try block without a throws clause in the method, it can be _safely_ re-thrown in the catch block.

> 
> I would prefer a `try...finally` inside `enqueue`. The reason is that `enqueue` knows where the underlying queue has been touched successfully, and where it wasn't (anything happened after `offer` returned successfully). So if it was touched successfully, should `semaphoreRead.release`, otherwise should `semaphoreWrite.release`.
> 
> ```
> private boolean enqueue(...) {
>    boolean offered = false;
>    try {
>       offered = super.offer(...);
>       ...
>    } finally {
>       (offered? semaphoreRead: semaphoreWrite).release();
>    }
> }
> ```

 Either way works for me, let me try this.